cmd/influxd/restore.go

```diff
@@ -0,0 +1,257 @@
+package main
+
+import (
+	"encoding/binary"
+	"flag"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"math/rand"
+	"net/url"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/boltdb/bolt"
+	"github.com/influxdb/influxdb"
+	"github.com/influxdb/influxdb/raft"
+)
+
+// RestoreCommand represents the program execution for "influxd restore".
+type RestoreCommand struct {
+	// The logger passed to the ticker during execution.
+	Logger *log.Logger
+
+	// Standard input/output, overridden for testing.
+	Stderr io.Writer
+}
+
+// NewRestoreCommand returns a new instance of RestoreCommand with default settings.
+func NewRestoreCommand() *RestoreCommand {
+	return &RestoreCommand{
+		Stderr: os.Stderr,
+	}
+}
+
+// Run excutes the program.
+func (cmd *RestoreCommand) Run(args ...string) error {
+	// Set up logger.
+	cmd.Logger = log.New(cmd.Stderr, "", log.LstdFlags)
+	cmd.Logger.Printf("influxdb restore, version %s, commit %s", version, commit)
+
+	// Parse command line arguments.
+	config, path, err := cmd.parseFlags(args)
+	if err != nil {
+		return err
+	}
+
+	// Remove broker & data directories.
+	if err := os.RemoveAll(config.BrokerDir()); err != nil {
+		return fmt.Errorf("remove broker dir: %s", err)
+	} else if err := os.RemoveAll(config.DataDir()); err != nil {
+		return fmt.Errorf("remove data dir: %s", err)
+	}
+
+	// Open snapshot file.
+	f, err := os.Open(path)
+	if err != nil {
+		return fmt.Errorf("open: %s", err)
+	}
+	defer f.Close()
+
+	// Create reader and extract manifest.
+	sr := influxdb.NewSnapshotReader(f)
+	ss, err := sr.Snapshot()
+	if err != nil {
+		return fmt.Errorf("snapshot: %s", err)
+	}
+
+	// Unpack snapshot files into data directory.
+	if err := cmd.unpack(config.DataDir(), sr); err != nil {
+		return fmt.Errorf("unpack: %s", err)
+	}
+
+	// Generate broker & raft directories from manifest.
+	if err := cmd.materialize(config.BrokerDir(), ss, config.BrokerURL()); err != nil {
+		return fmt.Errorf("materialize: %s", err)
+	}
+
+	// Notify user of completion.
+	cmd.Logger.Println("restore complete")
```

Comment:
It would be nice if this message explicitly printed the path of the read file, so the user can be clear that his path was picked up correctly by the command. E.g. "restore complete using file /var/opt/backup20150101.tar".